src/bokeh/core/property/container.py

```diff
@@ -202,11 +202,25 @@ def validate(self, value: Any, detail: bool = True) -> None:
 
         key_is_valid = self.keys_type.is_valid
         value_is_valid = self.values_type.is_valid
-        if isinstance(value, dict) and all(key_is_valid(key) and value_is_valid(val) for key, val in value.items()):
-            return
-
-        msg = "" if not detail else f"expected an element of {self}, got {value!r}"
-        raise ValueError(msg)
+        expected = f"expected a dict of type {self}"
+
+        if not isinstance(value, dict):
+            raise ValueError(f"{expected}, got a value of type {type(value)}" if detail else "")
+
+        bad_keys = [str(k) for k in value if not key_is_valid(k)]
+        bad_value_keys = [str(k) for (k, v) in value.items() if not value_is_valid(v)]
+        exception_header = f"{expected}, got a dict with"
+        bad_keys_str = f"invalid keys: {', '.join(bad_keys)}"
+        bad_key_values_str = f"invalid values for keys: {', '.join(bad_value_keys)}"
```

Comment:
My only comment is that this naming seems inconsistent, I'd expect `bad_value_keys -> bad_value_keys_str` based on `bad_keys -> bad_keys_str`

Reply:
Shoot, you are right. If you want to wait I can patch that when I get home.

Reply:
sure I'll merge when you give you 👍